internal/util/debug/debug.go

```diff
@@ -97,25 +123,98 @@ func Listen(opts *ListenOpts) (*Handler, error) {
 		}),
 	))
 
+	http.HandleFunc("/debug/archive", promhttp.InstrumentHandlerDuration(
+		archiveDurations,
+		http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+			rw.Header().Set("Content-Type", "application/zip")
+
+			rw.Header().Set(
+				"Content-Disposition",
+				fmt.Sprintf("attachment; filename=FerretDB-debug-%d.zip", time.Now().UnixMilli()),
+			)
+
+			ctx := req.Context()
+			zipWriter := zip.NewWriter(rw)
+
+			defer func() {
+				if err := zipWriter.Close(); err != nil {
+					opts.L.Error("Archive handler failed", zap.Error(err))
+					rw.WriteHeader(http.StatusInternalServerError)
+
+					return
+				}
+			}()
+
+			metricsFile, err := zipWriter.Create("metrics")
+			if err != nil {
+				opts.L.Error("Archive handler failed", zap.Error(err))
+				rw.WriteHeader(http.StatusInternalServerError)
+
+				return
+			}
+
+			// we use *http.Request instead of http.Get function to provide the ctx
+			scrapeReq := must.NotFail(http.NewRequestWithContext(
+				ctx,
+				http.MethodGet,
+				fmt.Sprintf("http://%s%s", req.Host, "/debug/metrics"),
+				nil,
+			))
+
+			resp, err := http.DefaultClient.Do(scrapeReq)
+			if err != nil {
+				opts.L.Error("Archive handler failed - metrics failed", zap.Error(err))
+				rw.WriteHeader(http.StatusInternalServerError)
+
+				return
+			}
+
+			_, err = io.Copy(metricsFile, resp.Body)
+
+			_ = resp.Body.Close()
+
+			if err != nil {
+				opts.L.Error("Archive handler failed", zap.Error(err))
+				rw.WriteHeader(http.StatusInternalServerError)
+
+				return
+			}
+
+			heapFile, err := zipWriter.Create("heap")
+			if err != nil {
+				opts.L.Error("Archive handler failed", zap.Error(err))
+				rw.WriteHeader(http.StatusInternalServerError)
+
+				return
+			}
+
+			scrapeReq.URL.Path = "/debug/pprof/heap"
+
+			resp, err = http.DefaultClient.Do(scrapeReq)
+			if err != nil {
+				opts.L.Error("Archive handler failed - pprof failed", zap.Error(err))
+				rw.WriteHeader(http.StatusInternalServerError)
+
+				return
+			}
+
+			_, err = io.Copy(heapFile, resp.Body)
+			_ = resp.Body.Close()
```

Comment:
Same here, error from Close should be logged

Reply:
https://github.com/FerretDB/FerretDB/pull/3895#discussion_r1684277509